File: hdf-diff/src/effective-checksum.ts
import { governingOverrideIndex, sha256, type StatusOverrideInput } from '@mitre/hdf-utilities';
import { computeEffectiveStatus } from './status.js';

interface ImpactOverrideLike {
  value: number;
}

interface OverrideLike {
  type?: string;
  status?: string;
  impact?: ImpactOverrideLike;
  appliedAt?: string;
  expiresAt: string;
}

export interface EffectiveChecksum {
  algorithm: 'sha256';
  value: string;
}

/** Maps overrides onto the shared neutral selection shape (applied/expiry
 * window only; eligibility is passed separately). */
function overrideWindows(overrides: readonly OverrideLike[]): StatusOverrideInput[] {
  return overrides.map((o) => ({ appliedAt: o.appliedAt, expiresAt: o.expiresAt }));
}

/**
 * Determine the effective impact of a requirement: the most recently applied
 * non-expired override carrying an impact value wins (the schema's definition
 * of effectiveImpact, selected by the shared governing helper); with no
 * overrides, a stored effectiveImpact field is honored; otherwise the base
 * impact. (Go parity: ComputeEffectiveImpact in hdf-diff/go/effective_checksum.go.)
 */
export function computeEffectiveImpact(
  requirement: Record<string, unknown>,
  referenceTimestamp?: string,
): number {
  const baseImpact = (requirement['impact'] as number | undefined) ?? 0;
  const overrides = requirement['statusOverrides'] as OverrideLike[] | undefined;

  if (overrides && overrides.length > 0) {
    const i = governingOverrideIndex(
      overrideWindows(overrides),
      (idx) => overrides[idx]?.impact !== undefined,
      referenceTimestamp,
    );
    if (i >= 0) {
      return overrides[i]!.impact!.value;
    }
    return baseImpact;
  }

  const effectiveImpact = requirement['effectiveImpact'] as number | undefined;
  if (effectiveImpact !== undefined) {
    return effectiveImpact;
  }
  return baseImpact;
}

/**
 * Return the Override_Type of the governing (most recently applied
 * non-expired) override, a stored disposition field when no overrides are
 * present, or null. (Go parity: ComputeDisposition.)
 */
export function computeDisposition(
  requirement: Record<string, unknown>,
  referenceTimestamp?: string,
): string | null {
  const overrides = requirement['statusOverrides'] as OverrideLike[] | undefined;

  if (overrides && overrides.length > 0) {
    const i = governingOverrideIndex(overrideWindows(overrides), () => true, referenceTimestamp);
    if (i >= 0) {
      return overrides[i]?.type ?? null;
    }
    return null;
  }

  const disposition = requirement['disposition'] as string | undefined;
  return disposition ?? null;
}

/**
 * Hash the resolved effective posture of a requirement: sha256 over the
 * canonical JSON {"status":<resolved>,"impact":<resolved>,"disposition":<type|null>}.
 * Flips exactly when the operative status, impact, or disposition changes;
 * stable under all other document churn. Byte-identical with the Go
 * implementation (ComputeEffectiveChecksum) — the key order of the canonical
 * object is part of the contract. Pass the document timestamp as
 * referenceTimestamp for determinism. A missing or unparseable reference
 * falls back to the wall clock inside the shared governing-override helper,
 * so callers that need reproducible output must always supply a valid
 * timestamp.
 */
export async function computeEffectiveChecksum(
  requirement: Record<string, unknown>,
  referenceTimestamp?: string,
): Promise<EffectiveChecksum> {
  const canonical = JSON.stringify({
    status: computeEffectiveStatus(requirement, referenceTimestamp),
    impact: computeEffectiveImpact(requirement, referenceTimestamp),
    disposition: computeDisposition(requirement, referenceTimestamp),
  });
  return { algorithm: 'sha256', value: await sha256(canonical) };
}

File: hdf-diff/go/effective_checksum.go
package diff

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	hdfutil "github.com/mitre/hdf-libs/hdf-utilities/go/v3"
)

// effectiveFields is the canonical hash input for ComputeEffectiveChecksum.
// Field order is the canonical serialization order; both language
// implementations must emit identical bytes.
type effectiveFields struct {
	Status      string  `json:"status"`
	Impact      float64 `json:"impact"`
	Disposition *string `json:"disposition"`
}

// overrideWindows maps schema overrides onto the shared neutral selection
// shape (applied/expiry window only; eligibility is passed separately).
func overrideWindows(overrides []hdf.StatusOverride) []hdfutil.StatusOverrideInput {
	windows := make([]hdfutil.StatusOverrideInput, len(overrides))
	for i := range overrides {
		windows[i] = hdfutil.StatusOverrideInput{
			AppliedAt: overrides[i].AppliedAt,
			ExpiresAt: overrides[i].ExpiresAt,
		}
	}
	return windows
}

// ComputeEffectiveImpact determines the effective impact of a requirement:
// the most recently applied non-expired override carrying an impact value
// wins (the schema's definition of effectiveImpact, selected by the shared
// governing helper); with no overrides, a stored effectiveImpact field is
// honored; otherwise the base impact.
func ComputeEffectiveImpact(req hdf.EvaluatedRequirement, referenceTimestamp string) float64 {
	if len(req.StatusOverrides) > 0 {
		ref := hdfutil.ParseTimestamp(referenceTimestamp)
		carriesImpact := func(i int) bool { return req.StatusOverrides[i].Impact != nil }
		if i := hdfutil.GoverningOverrideIndex(overrideWindows(req.StatusOverrides), carriesImpact, ref); i >= 0 {
			return req.StatusOverrides[i].Impact.Value
		}
		return req.Impact
	}
	if req.EffectiveImpact != nil {
		return *req.EffectiveImpact
	}
	return req.Impact
}

// ComputeDisposition returns the Override_Type of the governing (most
// recently applied non-expired) override, a stored disposition field when no
// overrides are present, or nil when nothing governs.
func ComputeDisposition(req hdf.EvaluatedRequirement, referenceTimestamp string) *hdf.OverrideType {
	if len(req.StatusOverrides) > 0 {
		ref := hdfutil.ParseTimestamp(referenceTimestamp)
		all := func(int) bool { return true }
		if i := hdfutil.GoverningOverrideIndex(overrideWindows(req.StatusOverrides), all, ref); i >= 0 {
			return &req.StatusOverrides[i].Type
		}
		return nil
	}
	return req.Disposition
}

// ComputeEffectiveChecksum hashes the resolved effective posture of a
// requirement: sha256 over the canonical JSON of
// {"status":<resolved>,"impact":<resolved>,"disposition":<type|null>}.
// It flips exactly when the requirement's operative status, impact, or
// disposition changes, and is stable under all other document churn
// (results detail, timestamps, tags). referenceTimestamp anchors override
// expiry — pass the document timestamp, never wall clock, for determinism.
func ComputeEffectiveChecksum(req hdf.EvaluatedRequirement, referenceTimestamp string) *hdf.Checksum {
	fields := effectiveFields{
		Status: ComputeEffectiveStatus(req, referenceTimestamp),
		Impact: ComputeEffectiveImpact(req, referenceTimestamp),
	}
	if d := ComputeDisposition(req, referenceTimestamp); d != nil {
		s := string(*d)
		fields.Disposition = &s
	}

	raw, err := json.Marshal(fields)
	if err != nil {
		return nil
	}
	hash := sha256.Sum256(raw)
	return &hdf.Checksum{
		Algorithm: hdf.Sha256,
		Value:     hex.EncodeToString(hash[:]),
	}
}

// StampEffectiveChecksums sets effectiveChecksum on every requirement of a
// map-decoded hdf-results document (baselines[].requirements and any
// top-level requirements array), leaving all other fields untouched. Used by
// document-rewriting tooling (hdf convert, amendment merge) that operates on
// maps to preserve unknown/extension fields.
func StampEffectiveChecksums(doc map[string]interface{}, referenceTimestamp string) error {
	stampList := func(reqsRaw interface{}) error {
		reqs, ok := reqsRaw.([]interface{})
		if !ok {
			return nil
		}
		for _, rRaw := range reqs {
			reqMap, ok := rRaw.(map[string]interface{})
			if !ok {
				continue
			}
			raw, err := json.Marshal(reqMap)
			if err != nil {
				return err
			}
			var typed hdf.EvaluatedRequirement
			if err := json.Unmarshal(raw, &typed); err != nil {
				// A requirement too malformed to type cannot be hashed; the
				// field is optional, so it stays unstamped rather than
				// failing the whole document (mirrors the merge tooling's
				// tolerance of malformed overrides).
				continue
			}
			cs := ComputeEffectiveChecksum(typed, referenceTimestamp)
			if cs == nil {
				continue
			}
			reqMap["effectiveChecksum"] = map[string]interface{}{
				"algorithm": string(cs.Algorithm),
				"value":     cs.Value,
			}
		}
		return nil
	}

	if baselinesRaw, ok := doc["baselines"].([]interface{}); ok {
		for _, bRaw := range baselinesRaw {
			baseline, ok := bRaw.(map[string]interface{})
			if !ok {
				continue
			}
			if err := stampList(baseline["requirements"]); err != nil {
				return err
			}
		}
	}
	return stampList(doc["requirements"])
}
